File: Lv1/DartGame.js
// 문제링크
// https://programmers.co.kr/learn/courses/30/lessons/17682

// 두자리 점수 계산을 배제한 풀이.. 수정 예정
// function solution(dartResult) {
//   var answer = 0;
//   let cnt = 0;
//   for (let idx = 0; idx < dartResult.length; idx++) {
//     if (isNaN(Number(dartResult[idx]))) {
//       if (dartResult[idx] === "S") {
//         answer += cnt;
//       } else if (dartResult[idx] === "D") {
//         answer += cnt ** 2;
//       } else if (dartResult[idx] === "T") {
//         answer += cnt ** 3;
//       } else if (dartResult[idx] === "*") {
//         answer *= 2;
//         cnt *= 2;
//       } else if (dartResult[idx] === "#") {
//         answer *= -1;
//         cnt *= -1;
//       }
//       cnt = 0;
//     } else {
//       cnt += Number(dartResult[idx]);
//     }
//   }

//   return answer;
// }

// 블로그를 보고 참고한 풀이
function solution(dartResult) {
  var answer = 0;
  let score = 0;
  let cnt = [];
  for (let i = 0; i < dartResult.length; i++) {
    //점수가 주어질때
    if (!isNaN(dartResult[i])) {
      //i-1이 1이고 i가 숫자면 10점.. //즉 10점인 경우와 그 외의 경우 처리
      score = Number(dartResult[i - 1]) === 1 ? 10 : Number(dartResult[i]);
      //보너스 S일때
    } else if (dartResult[i] === "S") {
      cnt.push(score);
      //보너스 D일때
    } else if (dartResult[i] === "D") {
      cnt.push(Math.pow(score, 2));
      //보너스 T일때
    } else if (dartResult[i] === "T") {
      cnt.push(Math.pow(score, 3));
      //옵션 *일떄
    } else if (dartResult[i] === "*") {
      cnt[cnt.length - 2] = cnt[cnt.length - 2] * 2;
      cnt[cnt.length - 1] = cnt[cnt.length - 1] * 2;
      //옵션 #일때
    } else if (dartResult[i] === "#") {
      cnt[cnt.length - 1] = -1 * cnt[cnt.length - 1];
    }
  } //3개의 점수 합산
  answer = cnt.reduce((acc, cur) => acc + cur, 0);
  return answer;
}

console.log(solution("1D2S#10S"));

File: Lv1/DartGame.py
# 문제링크
# https://programmers.co.kr/learn/courses/30/lessons/17682

def solution(dartResult):
    stack = [] # 다트의 점수들을 담아줄 배열
    dartResult = dartResult.replace("10", "A") # 10점은 두자리 수 이므로 A로 변환
    bonus = {'S': 1, 'D': 2, 'T': 3} # 싱글 더블 트리플 보너스 점수 (각 숫자만큼 제곱) 담아준 딕셔너리
    
    for i in dartResult: # 입력받은 점수와 보너스 문자열 순회
        if i.isdigit() or i=='A': # 순회하는 문자가 숫자거나 A(10)이라면,
            stack.append(10 if i == 'A' else int(i)) # A면 10을 추가, A가 아니면 해당 숫자 추가
        elif i in ('S', 'D', 'T'): # 순회하는 문자가 보너스 중 하나라면,
            num = stack.pop() # 가장 마지막에 들어간 숫자를 지워주고
            stack.append(num ** bonus[i]) # 해당 보너스의 숫자만큼 제곱해준다.
        elif i == '#': # "#"이라면, 가장 마지막 점수는 빼야하기 때문에
            stack[-1] *= -1 # -1를 곱해준다.
        elif i == '*': # "*"이라면,
            num = stack.pop() # 가장 마지막 점수가 두배가 되므로
            if len(stack): # 점수가 존재한다면,
                stack[-1] *= 2 # 해당 점수는 두배가 된다.
            stack.append(2 * num) # 얻은 점수 목록에 2배를 한 값을 추가
    return sum(stack) # 계산된 점수들의 합을 리턴